lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4584,7 +4629,155 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     errorOccurred = true;
   }
 
+  AllocStackInst *emitDifferentiableViewSubscript(
+      ApplyInst *ai, SILType elType, SILValue adjointArray, SILValue fnRef,
+      CanGenericSignature genericSig, int index) {
+    auto &ctx = builder.getASTContext();
+    auto astType = elType.getASTType();
+    auto literal = builder.createIntegerLiteral(
+        ai->getLoc(), SILType::getBuiltinIntegerType(64, ctx), index);
+    auto intType = SILType::getPrimitiveObjectType(
+        ctx.getIntDecl()->getDeclaredType()->getCanonicalType());
+    auto intStruct = builder.createStruct(
+        ai->getLoc(), intType, {literal});
+    AllocStackInst *subscriptBuffer = builder.createAllocStack(
+        ai->getLoc(), elType);
+    auto swiftModule = getModule().getSwiftModule();
+    auto diffProto =  ctx.getProtocol(KnownProtocolKind::Differentiable);
+    auto diffConf = swiftModule->lookupConformance(
+        astType, diffProto);
+    assert(diffConf.hasValue() && "Missing conformance to `Differentiable`");
+    auto addArithProto = ctx.getProtocol(KnownProtocolKind::AdditiveArithmetic);
+    auto addArithConf = swiftModule->lookupConformance(
+        astType, addArithProto);
+    assert(addArithConf.hasValue() &&
+           "Missing conformance to `AdditiveArithmetic`");
+    auto subMap = SubstitutionMap::get(
+        genericSig, {astType},
+        {*addArithConf, *diffConf});
+    auto subscriptApply = builder.createApply(
+        ai->getLoc(), fnRef, subMap,
+        {subscriptBuffer, intStruct, adjointArray});
+    return subscriptBuffer;
+  }
+
+  void accumulateDifferentiableViewSubscriptDirect(
+      ApplyInst *ai, SILType elType, StoreInst *si,
+      AllocStackInst *subscriptBuffer) {
+    auto astType = elType.getASTType();
+    auto newAdjValue = builder.createLoad(
+        ai->getLoc(), subscriptBuffer, getBufferLOQ(astType, getAdjoint()));
+    addAdjointValue(
+        si->getParent(), si->getSrc(),
+        makeConcreteAdjointValue(ValueWithCleanup(
+                newAdjValue, makeCleanup(newAdjValue, emitCleanup))));
+    builder.createDeallocStack(
```

Comment:
The same scope as the alloc would be `emitDifferentiableViewSubscript`, and the dealloc can't go there--I'm not sure it would be better to put it in the caller when the allocation isn't there.